### PR TITLE
images: build Arch Linux packages offline

### PR DESCRIPTION
**Pull request by jelly**
This changes the way we build Arch Linux packages for tests on our
images to allow for offline building. The standard way of building using
extra-x86_64-build first synchronises the Arch mirrors and fetches all
build dependencies when required. In this commit a build chroot is
pre-generated with the required build dependencies for cockpit
cockpit-podman and cockpit-machines. Using makechrootpkg to build the
package in the existing chroot should also speed up the build a bit as
this no longer runs checkpkg by default.

 * [x] image-refresh arch

### CODE REVIEW

**Comment by jelly** [comment]
Test it here https://github.com/cockpit-project/cockpit-machines/pull/626

**Comment by cockpituous** [comment]
image-refresh arch done: https://github.com/cockpit-project/bots/commits/image-refresh-arch-20220317-162828

**Comment by cockpituous** [comment]
image-refresh in progress on 4-cockpit-7.
Log: https://logs.cockpit-project.org/logs/image-refresh-3106-20220317-162828/

**Comment by cockpituous** [comment]
image-refresh arch done: https://github.com/cockpit-project/bots/commits/image-refresh-arch-20220317-164242

File: images/scripts/arch.setup [hunk 1]
```diff
@@ -106,6 +106,33 @@ echo "%wheel ALL=(ALL) ALL" > /etc/sudoers.d/90-cockpit-wheel
...
+npm \
+pcp \
+perl-json \
+perl-locale-po \
+xmlto \
+"
+
+# Create build chroot for cockpit projects for offline builds, the base chroot
+# is created in /var/tmp/archbuild/root
+mkdir /var/tmp/archbuild
```
Comment (martinpitt) [inline images/scripts/arch.setup:133]: Nitpick, but /var/tmp/ is prone to getting cleaned up automatically. Better use /var/lib/ or /var/cache/.

File: images/scripts/arch.setup [hunk 2]
```diff
@@ -106,6 +106,33 @@ echo "%wheel ALL=(ALL) ALL" > /etc/sudoers.d/90-cockpit-wheel
 useradd -c Builder builder
 echo "builder ALL = NOPASSWD: /usr/sbin/makechrootpkg, /usr/sbin/mkarchroot, /usr/bin/*-x86_64-build" > /etc/sudoers.d/90-archbuild
 
+BUILD_DEPS="\
```
Comment (martinpitt) [inline images/scripts/arch.setup:109]: Is there some way to tell it "install all build dependencies of cockpit"? Then we'd only have a small (possilbly empty) list of extra packages. By far not a blocker though, it doesn't change often.

**Review by martinpitt — CHANGES_REQUESTED** [review]
Tests fail with `==> ERROR: No chroot dir defined, or invalid path '/var/tmp/archbuild'`. Nevertheless, that's probably easy to fix -- I really appreciate seeing this, thanks for figuring out!

File: images/scripts/arch.setup (hunk 2 above)
Comment (jelly) [reply images/scripts/arch.setup:109]: We actually do record the makedepends now in the pacman database. The sync database (db pulled from the mirrors) is a simple tarball with packagename-$version/desc where `desc` is a simple file describing the package:

```
%FILENAME%
zstd-1.5.2-3-x86_64.pkg.tar.zst

%NAME%
zstd

%BASE%
zstd

%VERSION%
1.5.2-3

%DESC%
Zstandard - Fast real-time compression algorithm

%CSIZE%
671794

%ISIZE%
2487249

%MAKEDEPENDS%
gtest
```

**Comment by cockpituous** [comment]
image-refresh in progress on centosci-tasks-jzqvf.
Log: https://logs-https-frontdoor.apps.ocp.ci.centos.org/logs/image-refresh-3106-20220318-080202/

File: images/scripts/arch.setup (hunk 2 above)
Comment (jelly) [reply images/scripts/arch.setup:109]: Raised questions for the pacman team, it would be nice to have something like apt build-deps has.

**Comment by cockpituous** [comment]
Task failed: https://logs-https-frontdoor.apps.ocp.ci.centos.org/logs/image-refresh-3106-20220318-080202/

**Comment by jelly** [comment]
> Task failed: https://logs-https-frontdoor.apps.ocp.ci.centos.org/logs/image-refresh-3106-20220318-080202/

http 400, requires a retry.

**Comment by cockpituous** [comment]
image-refresh in progress on 1-ci-srv-04.
Log: https://logs.cockpit-project.org/logs/image-refresh-3106-20220318-084141/

**Comment by cockpituous** [comment]
image-refresh arch done: https://github.com/cockpit-project/bots/commits/image-refresh-arch-20220318-085656

**Comment by cockpituous** [comment]
image-refresh in progress on 4-cockpit-9.
Log: https://logs.cockpit-project.org/logs/image-refresh-3106-20220318-085656/

File: image-customize [hunk 3]
```diff
@@ -189,10 +189,7 @@ class BuildAction(ActionBase):
         machine.execute(f"su builder {opt_build_options} /var/tmp/mkbuild.sh")
 
         # build binaries
-        machine.execute("cd /var/tmp/build; su builder -c extra-x86_64-build", timeout=1800)
-
-        # install packages
-        machine.execute("pacman -U --noconfirm /var/tmp/build/*.pkg.tar.zst")
+        machine.execute("cd /var/tmp/build; makechrootpkg -r /var/lib/archbuild/cockpit -U builder -- -i", timeout=1800)
```
Comment (jelly) [inline image-customize:192]: -i tells makepkg to install the build packages, however this is run **in the chroot** not outside! *faceplam* :face_with_head_bandage:

Comment (jelly) [reply image-customize]: As alternative we could call `makepkg --packagelist` and use that to install (it needs to be run from the PKGBUILD directory) so we don't hardcode the package extension here

File: images/scripts/arch.setup (hunk 2 above)
Comment (jelly) [reply images/scripts/arch.setup:109]: So, I'll need to extend https://gitlab.archlinux.org/pacman/pacman/-/blob/master/src/pacman/util.c#L1145 so pacman can print the makedepends as well. Currently this is limited to a few formatting strings.

**Comment by jelly** [comment]
Offline tests running here for c-machines https://logs.cockpit-project.org/logs/pull-626-20220318-130109-9655f4a4-arch-bots-3106/log.html

**Comment by martinpitt** [comment]
I'm glad that the other failure is gone, and we are only left with this rather obvious [sudo message change](https://logs.cockpit-project.org/logs/pull-3106-20220318-142316-83898246-arch-cockpit-project-cockpit/log.html#233-2). Can you please send a cockpit PR which just relaxes the expected string? We don't care that much whether sudo reports failed logins :grin: 

xkcd

**Comment by jelly** [comment]
PR made here https://github.com/cockpit-project/cockpit/pull/17151 waiting on https://logs.cockpit-project.org/logs/pull-17151-20220318-154509-ecffe463-arch-bots-3106/log.html

**Comment by jelly** [comment]
After landing the sudo fix, the test now runs succesfully https://github.com/cockpit-project/cockpit-machines/pull/626

**Pull request merged**